File: Commander/History.cs
namespace Commander;

class History
{
    public void Set(string path)
    {
        if (history.Count == 0 || history[^1] != path)
        {
            history.Add(path);
            position = history.Count - 1;
        }
    }

    public string? Get(bool reverse)
        => reverse
            ? GetNext()
            : GetPrevious();

    string? GetPrevious()
        => position > 0
            ? history[--position]
            : null;

    string? GetNext()
        => position < history.Count - 1
            ? history[++position]
            : null;

    int position;
    readonly List<string> history = [];
}

File: src/history.ts
export interface History {
    set: (path: string) => void
    get: (reverse?: boolean) => string|null
}

export const initializeHistory = () => {
    const history: string[] = []
    let position = -1

    const set = (path: string) => {
        if (history[history.length -1] != path) {
            history.push(path)
            position = history.length - 1
        }
    }

    const getPrevious = () => 
        position > 0
        ? history[--position]
        : null    
    
    const getReverse = () => 
        position < history.length - 1
        ? history[++position]
        : null    
        
    const get = (reverse?: boolean) => 
        reverse
        ? getReverse()
        : getPrevious()

    return {
        set,
        get
    } as History
}